### ISSUE TITLE
New rule suggestion: no-require-extension

### ISSUE THREAD

**Issue by denis-sokolov** [eslint/eslint#2542]
I would like to add a new rule to eslint - `no-require-extension`.
It would disallow adding an extension to a filename in a `require` call.
Perhaps a configurable list of extensions that need to be left out. In my use case it's only `.js` that needs to be left out.

The following would be invalid when the rule is enabled:
```javascript
require('./foo.js');
require('./foo/bar.js');
require('../foo.js');
```

The following would still be valid when the rule is enabled:
```javascript
require('./foo');
require('./foo/bar');
require('../foo');
require('foo/bar.js');
```

The use case is - there is a common programming pattern where you start with small and simple implementations and you can later grow them as needed. What started as `foo.js` can grow too long and need its own subdirectory to be split off into many files. This transition is smooth if everybody who depends on `foo` only used `require('./foo')`, but if dependents used `require('./foo.js')`, then either dependents need to be modified, or a stub `foo.js` needs to remain.

The error thus would prevent a maintainability error by a developer.

Having this error in the list will not only allow projects with this structure to enforce it, but perhaps having it in the list of rules can cause new developers to consider such a possibility for their projects.

The rule is very specific to a particular workflow, and definitely should be off by default (this choice is irrelevant with 1.0.0 using `--reset`).

**Comment by nzakas** [comment]
Why would this be allowed?

```
require('foo/bar.js');
```

**Comment by denis-sokolov** [comment]
The author of the code has no control over `foo`, that being a submodule.
The advantage of being able to switch a `foo.js` to ` foo/index.js` does not apply for 3rd party code.
In addition, the  author has no control over 3rd party code and API and that submodule could have `foo/bar.js` and `foo/bar/index.js`, for example.
Overall, my reasoning was to better be safe - if a linting rule has a false negative, not a big deal, it's as is if it wasn't there. A false positive, however, is a serious annoyance.

**Comment by mathieumg** [comment]
:+1: I'd welcome this rule too.

**Comment by nzakas** [comment]
@denis-sokolov hmm, I'm not buying that line of reasoning. `foo/bar.js` is the same as `foo/bar`. If a package author created both `foo/bar.js` and `foo/bar/index.js`, then that's probably a mistake.

I'd be in favor of this rule without the exception.

**Comment by denis-sokolov** [comment]
Sure, no problem. I was being conservative, but I'm glad to have it without the exception.
Am I to understand this is green light for writing a PR or is this still under consideration?

**Comment by ilyavolodin** [comment]
You should hold off with PR until the issue is going to be marked with "accepted" tag.

**Comment by nzakas** [comment]
We should consider if we want this as a separate rule or as an option to another rule #2539. Kind of seems like all of these suggestions are some form of "filter the string value passed to `require`" so perhaps are really just one rule?

**Comment by denis-sokolov** [comment]
This makes sense.

**Comment by denis-sokolov** [comment]
So far this seems clearly convenient as an option in #2555.
